### PR TITLE
[Tools] Create bootstrap (dependency installation) script for Ubuntu environments

### PR DESCRIPTION
**Pull request by johnsaigle**
### Brief summary of changes

This PR introduces a new tools script to "bootstrap" a production environment of LORIS. This is intended as a post-install step to ensure that a developer's system has the sufficient software required to run LORIS. 

This is done by:
* Checking that the PHP version is sufficient
* Checking that the Apache version is sufficient
* Checking that miscellaneous required packages are present

Note that this is not intended to be used on production environments. A separate solution involving proper package managers is more suitable. This only makes a developer's life a little easier when getting started.

There are two other elements of this PR:

1. I've used an OO approach and created an interface for future bootstrappers for other environments to be added. CentOS probably makes sense to create as we use that frequently.
2. There is a small CLI_Helper class that does generic shell-like things. In the future this can be used like the `Utility` class except it will contain functions commonly used by tools/ scripts.

### Testing
1. Go to the tools/ directory and run `php bootstrap.php`. 
2. If you have everything installed it will exit and say "All requirements satisfied". Try doing `sudo apt remove $package` and running the script again.
3. If you are missing a package the script will tell you which are missing.
4. If you run the script with `--install` then the script will attempt to install the missing packages.
5. Repeat the above steps but use `make bootstrap` or `make dev` instead of running the script directly.

### Related

This is a rewrite of #3981

### CODE REVIEW

**Comment by johnsaigle** [comment]
TODO:
- [ ] Should work from outside `tools/` dir
- [ ] Should be included in `make dev` script
- [ ] Script should terminate if not on Ubuntu or Debian

**Comment by xlecours** [comment]
This seems to add a lot of maintenance overhead. What does it aim to prevent/reduce?

**Comment by johnsaigle** [comment]
The idea is to provide a step to make sure that a developer's system is ready to use LORIS. Just as we have composer, npm, etc. to help us manage our dependencies, this integrates with those tools to help get things set-up.

As an example when we upgrade PHP versions there are usually a lot of people wondering why "composer doesn't work" as it is usually telling them that they are missing e.g. the newest PHP 7.X extension to work with MySQL.

This script takes care of this. It's a part of `make dev` so nobody will have to manually install all these apt packages when we upgrade.

It also automates the following steps that are buried in the documentation and are a source of user error especially when new developers come into the lab. Note that there is already existing overhead in maintaining this documentation that this PR helps alleviate.
https://github.com/aces/Loris/wiki/Installing-Loris-in-Depth#getting-prerequisites

In addition it would help with issues like the following we've had in the past:
* https://github.com/aces/Loris/pull/3204
* https://github.com/aces/Loris/issues/3972

The overhead is there for two reasons:
1) To make this extensible to systems other than Ubuntu
2) The `CLI_Helper` is handy and can be used and added to during eventual refactoring of other `tools/` scripts.

**Comment by johnsaigle** [comment]
@PapillonMcGill I chose to write this script in PHP because most of our team knows php, not bash. This will have a better chance of being maintained if people can understand it. I had meetings with @driusan and Samir last year to talk this point through. Fortunately I'm pretty sure most Linux distros come with some PHP installed and this will tell them if it needs to be updated. 

I used the interface and OO structure for the script so that a CentOS class can be created at some point. The author can use the Ubuntu class as a guide for a `CentOSBootstrapper` class by implementing a few simple functions. A case can be added in the switch-case in the `bootstrap` script to use it afterward.

**Comment by PapillonMcGill** [comment]
@johnsaigle Thanks for the update on the requirements

**Comment by PapillonMcGill** [comment]
LGTM, waiting Travis

**Comment by johnsaigle** [comment]
Need to add `make` and `npm` to this list. Maybe composer also.

**Comment by kchatpar** [comment]
@johnsaigle The --install flag is not working, here's the error I'm getting:

lorisadmin@kchatpar-dev:/var/www/loris/tools$ php bootstrap.php --install
dpkg-query: package ‘php-json’ is not installed and no information is available
Use dpkg --info (= dpkg-deb --info) to examine archive files,
and dpkg --contents (= dpkg-deb --contents) to list their contents.

[-] ERROR: Command `dpkg -s php-json` failed (error code 1)
dpkg-query: package ‘npm’ is not installed and no information is available
Use dpkg --info (= dpkg-deb --info) to examine archive files,
and dpkg --contents (= dpkg-deb --contents) to list their contents.

[-] ERROR: Command `dpkg -s npm` failed (error code 1)
Found missing packages required by LORIS:
       - php-json
       - npm
These may be installed by running this script with the --install flag.

**Comment by johnsaigle** [comment]
@kchatpar Thanks for testing. I fixed the bug and packages should install now. Could you re-test?

File: tools/CLI_Helper.class.inc [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        if (is_iterable($output)) {
+            foreach ($output as $item) {
+                $error[] = $item . PHP_EOL;
+            }
+        }
+        return $error;
+    }
+
+    /**
+     * Returns the OS distribution based on the /etc/os-release file.
```
[the remark this answers is not in the document]
Comment (PapillonMcGill) [reply tools/CLI_Helper.class.inc]: unfixed!
```suggestion
     * Returns the OS distribution based on the hostnamectl tools.
```

File: tools/Bootstrapper.class.inc [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ * @author   John Saigle <john.saigle@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+/**
+ * Defines an interface to be extended by classes corresponding to different
+ * UNIX-like distributions. LORIS requires a minimum PHP and Apache version as
+ * well as various system tools and PHP extensions in order to function. As
+ * different systems will have differnt package managers and other programs used
```
[the remark this answers is not in the document]
Comment (PapillonMcGill) [reply tools/Bootstrapper.class.inc]: unfixed!
```suggestion
 * different systems will have different package managers and other programs used
```

**Comment by kchatpar** [comment]
@johnsaigle I re-tested the line: php bootstrap.php --install and I'm still getting errors. Here's my stacktrace:

lorisadmin@kchatpar-dev:/var/www/loris/tools$ php bootstrap.php --install
dpkg-query: package 'php-json' is not installed and no information is available
Use dpkg --info (= dpkg-deb --info) to examine archive files,
and dpkg --contents (= dpkg-deb --contents) to list their contents.

[-] ERROR: Command `dpkg -s php-json` failed (error code 1)
dpkg-query: package 'npm' is not installed and no information is available
Use dpkg --info (= dpkg-deb --info) to examine archive files,
and dpkg --contents (= dpkg-deb --contents) to list their contents.

[-] ERROR: Command `dpkg -s npm` failed (error code 1)
[*] Installing requirements...
[sudo] password for lorisadmin: 
E: Unable to correct problems, you have held broken packages.

PHP Fatal error:  Uncaught Error: [] operator not supported for strings in /var/www/loris/tools/CLI_Helper.class.inc:99
Stack trace:
#0 /var/www/loris/tools/CLI_Helper.class.inc(73): CLI_Helper->bashErrorToString('sudo apt-get in...', Array, 100)
#1 /var/www/loris/tools/UbuntuBootstrapper.php(171): CLI_Helper->doExec('sudo apt-get in...')
#2 /var/www/loris/tools/UbuntuBootstrapper.php(152): UbuntuBootstrapper->installPackage('npm')
#3 /var/www/loris/tools/bootstrap.php(69): UbuntuBootstrapper->installPackages(Array)
#4 {main}
  thrown in /var/www/loris/tools/CLI_Helper.class.inc on line 99

**Comment by johnsaigle** [comment]
@kchatpar Thanks for testing. I've fixed the bug you found. Could you test again?

**Comment by kchatpar** [comment]
@johnsaigle I retested your script. Php bootstrap.php detected that npm was missing on my VM. The ---install flag failed to install npm because it had some dependencies. If your script could detect which dependencies are needed to be installed to successfully install the missing packages, this would be very useful. When a dependency is detected it could ask the user if they want to install the dependencies.

It would also be useful to output the outcome of all the checks. Example if PHP is up to date the user would get a message saying PHP version is installed and up to date.

**Comment by johnsaigle** [comment]
@kchatpar You can run the script with a `-v` flag and it will print the status of all installed packages.

I'm having trouble uninstalling `npm` in order to confirm your issue. Can you paste the error message you got?

**Comment by johnsaigle** [comment]
I added the `-f` flag in the latest commit as it should help resolve dependencies.

The script just calls out to `apt` which can built-in logic to install dependencies. That logic can be pretty complicated so I wouldn't want to try to implement recursive package installation functionality here. Hopefully the above flag fixes the problem.

**Comment by AlexandraLivadas** [comment]
Hello @johnsaigle 
I ran your script with the --install flag and then it became stuck at "Installing Requirements". When I re-ran it without the --install flag, I saw that it had managed to install all the packages that I had been missing (make, php-json) except for npm.

I then ran it again with --install and -v, and it seems to get stuck at "[+] Executing command `sudo apt-get -f install 'npm'`..."

Here is the output when I ran the script with "--install -v":
[+] Executing command `dpkg -s curl`...
[+] OK.
[+] Executing command `dpkg -s wget`...
[+] OK.
[+] Executing command `dpkg -s zip`...
[+] OK.
[+] Executing command `dpkg -s unzip`...
[+] OK.
[+] Executing command `dpkg -s php-json`...
[+] OK.
[+] Executing command `dpkg -s make`...
[+] OK.
[+] Executing command `dpkg -s npm`...
dpkg-query: package 'npm' is not installed and no information is available
Use dpkg --info (= dpkg-deb --info) to examine archive files,
and dpkg --contents (= dpkg-deb --contents) to list their contents.

[-] ERROR: Command `dpkg -s npm` failed (error code 1)
[+] Executing command `dpkg -s software-properties-common`...
[+] OK.
[+] Executing command `dpkg -s php-ast`...
[+] OK.
[+] Executing command `dpkg -s php7.2`...
[+] OK.
[+] Executing command `dpkg -s php7.2-mysql`...
[+] OK.
[+] Executing command `dpkg -s php7.2-xml`...
[+] OK.
[+] Executing command `dpkg -s php7.2-json`...
[+] OK.
[+] Executing command `dpkg -s php7.2-mbstring`...
[+] OK.
[+] Executing command `dpkg -s php7.2-gd`...
[+] OK.
[+] Executing command `dpkg -s libapache2-mod-php7.2`...
[+] OK.
[*] Installing requirements...
[+] Executing command `sudo apt-get -f install 'npm'`...

**Comment by johnsaigle** [comment]
Thanks for testing @AlexandraLivadas. I'll have to look into whether `npm` can be installed this way.

**Comment by johnsaigle** [comment]
@AlexandraLivadas I wasn't able to reproduce this when running the script on my VM.

I tried:
`sudo apt remove npm`
`php bootstrap.php`
`php bootstrap.php --install`

which succeeded. Could you try again and see if this works for you?

**Comment by johnsaigle** [comment]
The solution to the `npm` issue is posted in #4172 and I'll amend this script to use it.

**Comment by christinerogers** [comment]
re Calling this script `bootstrap` : could we consider calling it something like "Verify/Fix Environment" ?  because: 
* dependency/package confusion  with actual Bootstrap
* Let's describe what it does (check/fix dependencies)

**Comment by johnsaigle** [comment]
Bootstrapping is a software development term that means exactly that. It's probably where Bootstrap gets its name.

This is the original inspiration for the script and where the name came from:
https://github.com/github/scripts-to-rule-them-all#scriptbootstrap

> script/bootstrap is used solely for fulfilling dependencies of the project.
This can mean RubyGems, npm packages, Homebrew packages, Ruby versions, Git submodules, etc.
**The goal is to make sure all required dependencies are installed.**

**Comment by johnsaigle** [comment]
I've updated the script to check for a NodeJS dependency in a different way than before. The script no longer attempts to install it for the user but will warn if they don't have it.

File: tools/UbuntuBootstrapper.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        // When preg_match returns 0 it means no match was found.
+        return preg_match($pattern, $apacheVersion) !== 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return bool Whether the requirement is met.
+     */
+    public function nodeJSRequirementSatisfied(): bool
```
Comment (PapillonMcGill) [inline tools/UbuntuBootstrapper.php]: This function is never called.

Comment (johnsaigle) [reply tools/UbuntuBootstrapper.php]: Thanks, the latest commit should fix this.

**Comment by johnsaigle** [comment]
@ridz1208 @driusan Could you please provide feedback on this PR? The issue of system dependencies keeps coming up in bug reports and in documentation updates.

Even if this isn't the approach we end up going with, this issue should move forward.

**Pull request closed without merging**